---
title: "Differential Abundance"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	message = FALSE,
	warning = FALSE,
	echo = FALSE,
	cache = FALSE
)

library(data.table)
library(DT)
library(rtracklayer)

source("scripts/ipsmodel.R")

PLOT.FDR.THRESHOLD <- 0.05

# Convert a p-value into a significance marker
sig.stars <- function(p) {
  pp <- ifelse((p > 0) & (p < 1), p, 0.5)
  ifelse(!is.finite(p) | (p < 0) | (p > 1), "?",
         ifelse(p > 0.05, " ",
                ifelse(p == 0, "*****",
                       strrep("*", pmin(ceiling(-log10(pp)*(1+.Machine$double.eps))-1, 5)))))
}
```

```{r load}
# Load list of knockouts, find assumed neutral set
message("Loading knockouts from ", snakemake@input$gff)
knockouts <- readGFF(snakemake@input$gff, version=2)
NEUTRAL <- knockouts$Name[knockouts$Neutral == 1]

# Load UMI count tables produced by TRUmiCount
load.trumicount.output <- function(file) {
  message("Loading TRUmiCount results from ", file)
  t <- data.table(read.table(file, header=TRUE))
  # Make sure there are exactly two rows per knockout, plus one for 'ambiguous' and 'unmatched'
  allgenes <- data.table(gene=c(paste0(knockouts$Name, ':5p'),
                                paste0(knockouts$Name, ':3p'),
                                'ambiguous', 'unmatched'))
  t <- t[allgenes,, on="gene"]
  # The raw "gene" column contains <gene>:<flank>, split into two separate fields
  t[, flank := sapply(strsplit(as.character(gene), split=":", fixed=TRUE), FUN=function(f) { f[2] }) ]
  t[, knockout := sapply(strsplit(as.character(gene), split=":", fixed=TRUE), FUN=function(f) { f[1] }) ]
}
counts.flank.in <- load.trumicount.output(snakemake@input[["pool_in"]])
counts.flank.out <- load.trumicount.output(snakemake@input[["pool_out"]])

# Aggregate the data for the 5' and 3' flank of each gene.
# We add the counts and average the losses of the two flanks for each gene,
# see docs/pipeline_description.pdf for details
combine.flanks <- function(counts.flank) {
  counts.flank[!(knockout %in% c('ambiguous', 'unmatched')), list(
    n = sum(n.obs, na.rm=TRUE),
    loss = mean(loss, na.rm=TRUE)
  ), by="knockout"]
}
counts.in <- combine.flanks(counts.flank.in)
counts.out <- combine.flanks(counts.flank.out)

# Combine input and output pool data into a single table
counts <- merge(counts.in, counts.out, by="knockout", suffixes=c(".in", ".out"))[,list(
  knockout,
  is.neutral=knockout %in% NEUTRAL,
  n.in, loss.in,
  n.out, loss.out,
  abundance.in=n.in / (2 * (1-loss.in)),
  abundance.out=n.out / (2 * (1-loss.out))
)]
```

# 1. Files

```{r}
knitr::kable(data.table(kind=c("knockouts",
                               "input pool (pre-infection) knockout abundances ",
                               "output pool (post-infection) knockout abundances",
                               "log2 fold changes & significance"),
                        type=c("read", "read", "read", "written"),
                        file=c(snakemake@input$gff,
                               snakemake@input$pool_in,
                               snakemake@input$pool_out,
                               snakemake@output$table),
                        report=c("",
                                 paste0("[TRUmiCount report](", basename(snakemake@input$trumicount_pdf_in), ")"),
                                 paste0("[TRUmiCount report](", basename(snakemake@input$trumicount_pdf_out), ")"),
                                 "")))
```

# 2. Quality Control

## Read and UMI count statistics

TODO

For the distribution of the number of reads per UMI, see the *TRUmiCount reports* linked in the list of files above.

## Correlation of 3' and 5' Flank

```{r table_corr_flanks}
group.by.knockout <- function(data) {
  data[, list(n5p=n.obs[flank=="5p"],
              l5p=loss[flank=="5p"],
              n3p=n.obs[flank=="3p"],
              l3p=loss[flank=="3p"])
       , by="knockout"]
}

knitr::kable(data.table(
  `Pool`=c(
  "input",
  "output",
  "input",
  "output"
  ),
  Type=c(
  "Raw",
  "Raw",
  "Loss-corr.",
  "Loss-corr."
  ),
  `Correlation 5' vs. 3'`=c(
  group.by.knockout(counts.flank.in)[, cor(n5p, n3p, use='complete.obs')],
  group.by.knockout(counts.flank.out)[, cor(n5p, n3p, use='complete.obs')],
  group.by.knockout(counts.flank.in)[, cor(n5p/(1-l5p), n3p/(1-l3p), use='complete.obs')],
  group.by.knockout(counts.flank.out)[, cor(n5p/(1-l5p), n3p/(1-l3p), use='complete.obs')])
), digits=3)
```

```{r plot_corr_flanks, fig.height=8, fig.width=8}
p <- function(n.obs.5p, n.obs.3p, type, ...) {
  plot.new()
  plot.window(xlim=range(n.obs.5p[n.obs.5p>0], na.rm=TRUE),
              ylim=range(n.obs.3p[n.obs.3p>0], na.rm=TRUE), log="xy")
  axis(side=1)
  axis(side=2)
  title(xlab=paste0(type, " #UMIs on 5' flank of knockout"),
       ylab=paste0(type, " #UMIs on 3' flank of knockout"),
       ...)
  points(ifelse(n.obs.5p > 0, n.obs.5p, 10**(par("usr")[1])),
         ifelse(n.obs.3p > 0, n.obs.3p, 10**(par("usr")[3])),
         pch=ifelse((n.obs.5p == 0) | (n.obs.3p == 0), 4, 1),
         xpd=NA)
}

layout(cbind(c(1,3), c(2,4)))
d <- group.by.knockout(counts.flank.in)[, {
  p(n5p, n3p, main="Input 5' vs. 3' flank (raw)", type="Raw")
}]
d <- group.by.knockout(counts.flank.out)[, {
  p(n5p, n3p, main="Output 5' vs. 3' flank (raw)", type="Raw")
}]
d <- group.by.knockout(counts.flank.in)[, {
  p(n5p/(1-l5p), n3p/(1-l3p), main="Input 5' vs. 3' flank (loss-corrected)", type="Loss-corr.")
}]
d <- group.by.knockout(counts.flank.out)[, {
  p(n5p/(1-l5p), n3p/(1-l3p), main="Output 5' vs. 3' flank (loss-corrected)", type="Loss-corr.")
}]
```

## Correlation of Input and Output

```{r table_corr_io}
group.by.knockout <- function(data) {
  data[, list(n5p=n.obs[flank=="5p"],
              l5p=loss[flank=="5p"],
              n3p=n.obs[flank=="3p"],
              l3p=loss[flank=="3p"])
       , by="knockout"]
}

knitr::kable(data.table(
  Type=c(
  "Raw",
  "Loss-corr."
  ),
  `Correlation Input vs. Output`=c(
  counts[, cor(n.in, n.out, use='complete.obs')],
  counts[, cor(abundance.in, abundance.out, use='complete.obs')])
), digits=3)
```

```{r plot_corr_io, fig.height=4, fig.width=8}
p <- function(abd.in, abd.out, type, ...) {
  plot.new()
  plot.window(xlim=range(abd.in[abd.in>0], na.rm=TRUE),
              ylim=range(abd.out[abd.out>0], na.rm=TRUE), log="xy")
  axis(side=1)
  axis(side=2)
  title(xlab=paste0(type, " #UMIs in the input"),
       ylab=paste0(type, " #UMIs in the output"),
       ...)
  points(ifelse(abd.in > 0, abd.in, 10**(par("usr")[1])),
         ifelse(abd.out > 0, abd.out, 10**(par("usr")[3])),
         pch=ifelse((abd.in == 0) | (abd.out == 0), 4, 1),
         xpd=NA)
}

layout(cbind(1, 2))
d <- counts[, {
  p(n.in, n.out, main="Input vs. Output", type="Raw")
}]
d <- counts[, {
  p(abundance.in, abundance.out, main="Input vs. Output (loss-corrected)", type="Loss-corr.")
}]
```

## Model Parameter & Fit

```{r fit_model}
# Fit null model
null.model <- counts[is.neutral & (n.in > 0) & (n.out > 0), {
  ipsmodel.fit(n.out=n.out, n.in=n.in, l.out=loss.out, l.in=loss.in)
}]
knitr::kable(signif(null.model, digits=3),
             col.names=c('Scale (lambda)', 'Dispersion (d)'))
```

# 3. Results

## Log2 fold changes relative to neutral knockouts

```{r sig_test}
# Compute log2 fold changes and p-values
results <- counts[, list(
  knockout, is.neutral,
  n.out, loss.out, abundance.out,
  n.in, loss.in, abundance.in,
  log2fc=ifelse(is.finite(abundance.out) & is.finite(abundance.in) & (abundance.in>0) & (abundance.out>0),
                log2(abundance.out / (null.model$scale * abundance.in)),
                NA_real_),
  low.pval=ifelse(is.finite(n.out) & is.finite(n.in) & (n.out > 0) & (n.in > 0),
                  pipsmodel(n.out, null.model$scale, null.model$disp, n.in, loss.out, loss.in),
                  NA_real_),
  high.pval=ifelse(is.finite(n.out) & is.finite(n.in) & (n.out > 0) & (n.in > 0),
                  1 - pipsmodel(n.out - 1, null.model$scale, null.model$disp, n.in, loss.out, loss.in),
                  NA_real_)
)]

# Compute q-values using the method of Benjamini & Hochberg (BH)
results[, low.qval := p.adjust(low.pval, method='BH')]
results[, high.qval := p.adjust(high.pval, method='BH')]

# Write output table
write.table(results, file=snakemake@output$table, col.names=TRUE, sep="\t", row.names=FALSE)
```

```{r plot_log2fc}
# Create plot
par(mar=c(4,4,1,1))
plot.new()
xmax <- 10**ceiling(max(log10(results$abundance.in), na.rm=TRUE))
plot.window(xlim=c(1, xmax),
            ylim=range(results$log2fc, -results$log2fc, na.rm=TRUE),
            log="x")

# Plot (approximate) insigificance region
loss.out <- mean(results$loss.out, na.rm=TRUE)
loss.in <- mean(results$loss.in, na.rm=TRUE)
a.in <- 10**seq(from=0, to=log10(xmax), length.out=100)
l2fc.low <- log2(pmax(qipsmodel(p=rep(PLOT.FDR.THRESHOLD, along.with=a.in),
                                scale=null.model$scale, disp=null.model$disp,
                                n.in=a.in * (1 - loss.in), l.out=loss.out, l.in=loss.in) - 1, 0) /
                 ((1 - loss.out) * a.in * null.model$scale))
l2fc.high <- log2(pmax(qipsmodel(p=rep(1 - PLOT.FDR.THRESHOLD, along.with=a.in),
                                 scale=null.model$scale, disp=null.model$disp,
                                 n.in=a.in * (1 - loss.in), l.out=loss.out, l.in=loss.in) + 1, 0) /
                 ((1 - loss.out) * a.in * null.model$scale))
polygon(c(a.in, rev(a.in)),
        pmax(c(rev(cummin(rev(l2fc.low))),
               cummax(rev(l2fc.high))), par("usr")[3]-1),
        border=NA, col="lightgrey", xpd=FALSE)

# Plot points
d <- results[abundance.in > 0][order(is.neutral, low.qval <= PLOT.FDR.THRESHOLD, high.qval <= PLOT.FDR.THRESHOLD), {
  points(abundance.in,
         ifelse(abundance.out == 0, par("usr")[3], log2fc),
         xpd=NA,
         pch=ifelse((abundance.in == 0) | (abundance.out == 0), 4, 1),
         col=ifelse(is.neutral, 'darkgreen',
                    ifelse(low.qval <= PLOT.FDR.THRESHOLD, 'red',
                           ifelse(high.qval <= PLOT.FDR.THRESHOLD, 'blue',
                                  'darkgrey'))))

# Axis
axis(side=1)
axis(side=2)
title(xlab="Input Abundance", ylab=expression(paste(log[2], " fold change in output vs. neutral")))
}]
```

## Depletion of knockouts in the output compared to neutral knockout

```{r table_depleted}
DT::datatable(results[!is.na(log2fc), list(
  knockout, is.neutral, log2fc, pval=low.pval, qval=low.qval, sig=sig.stars(low.qval)
)], filter="top") %>%
  formatRound(columns=c("log2fc"), digits=2) %>%
  formatSignif(columns=c("pval", "qval"), digits=2)
```

## Enrichment of knockouts in the output compared to neutral knockout

```{r table_enriched}
DT::datatable(results[!is.na(log2fc), list(
  knockout, is.neutral, log2fc, pval=high.pval, qval=high.qval, sig=sig.stars(high.qval)
)], filter="top") %>%
  formatRound(columns=c("log2fc"), digits=2) %>%
  formatSignif(columns=c("pval", "qval"), digits=2)
```

